modules/monitoring/models/scheduledate.php

```diff
@@ -1,479 +1,485 @@
-		                    $exc_day_e = $day_range[1];
-		                }else{
-		                    $exc_day_e = $day_range[0];
-		                }
-		            }
-	                $strt_d = explode('-',$exc_day_s);
-	                $sy = (int)$strt_d[0];
-	                $sm = (int)$strt_d[1];
-	                $smn = $months[$sm-1];
-	                $sd = (int)$strt_d[2];
-	                $exc_start_day = strtotime("$sd $smn $sy");
-	                $end_d = explode('-',$exc_day_e);
-	                $ey = (int)$end_d[0];
-	                $em = (int)$end_d[1];
-	                $emn = $months[$em-1];
-	                $ed = (int)$end_d[2];
-	                $exc_end_day = strtotime("$ed $emn $ey");
-	                if($timestamp >= $exc_start_day && $timestamp <= $exc_end_day){
-	                    continue;
-	                }
-		        }
-		    }
-			if($ends_on && $ends_on != ''){
-				$ends_on = explode('-',$ends_on);
-				$ey = (int)$ends_on[0];
-				$em = (int)$ends_on[1];
-				$emn = $months[$em-1];
-				$ed = (int)$ends_on[2];
-				$end_on_timestamp = strtotime("$ed $emn $ey");
-				if($timestamp > $end_on_timestamp){
+
+				return false;
+			case 'month':
+				/**
+				 * Repeats every Nth month
+				 * Conditions:
+				 * 1) Targetdate's /month/ matches the scheduled month-interval
+				 * 2) Targetdate is the "<ordinal> <day> of <month>"
+				 */
+				return $downtime->match_month_interval($target_date) && $downtime->match_day_of_month($target_date);
+			case 'year':
+				/**
+				 * Repeats every Nth year
+				 * Conditions:
+				 * 1) Targetdate's /year/ matches the scheduled year-interval
+				 * 2) Targetdate's month matches recurring month
+				 * 3) Targetdate is the "<ordinal> <day> of <month>"
+				 */
+				return $downtime->match_year_interval($target_date) && $downtime->match_day_of_month($target_date);
+			default:
+				$msg = "Invalid recurrence: $unit";
+				$this->logger->log('error', $msg);
+				throw new UnexpectedValueException($msg);
+		}
+	}
+
+	/**
+	 * Schedule downtime
+	 *
+	 * @param $timestamp false|string Timestamp instead of the default target date (tomorrow)
+	 * @param $schedules false|array List of schedules instead of obtaining schedules from the database
+	 * @return int 1 if one or more ext commands failed, otherwise 0
+	 * @throws Exception
+	 */
+	public function schedule_downtime($timestamp = false, $schedules = false) {
+		// Store results in an array for processing later
+		$results = array();
+
+		// Get schedules from DB
+		if(!$schedules) {
+			$schedules = RecurringDowntimePool_Model::all();
+		}
+
+		// Create datetime object from argument, if provided, otherwise set to tomorrow
+		$target_date = new NinjaDateTime($timestamp ? '@' . $timestamp : 'tomorrow');
+
+		foreach ($schedules as $sched) {
+			$comment = $sched->get_comment();
+
+			try {
+				$downtime = new RecurringDowntime($sched);
+
+				if($downtime->recurrence_ends <= $target_date) {
```

Comment:
I assume the recurring downtimes are stored in the database somewhere, and we at L182 gets them all? I.e. all recurring scheduled downtimes ever created? Might it be worth only fetching those that are not expired ? Should be possible with a SQL statement to filter those out.

Reply:
Correct - and that's a good suggestion. I'm not familiar with the Kohana ORM but I'm guessing it shouldn't be too hard to implement.
Side effect: this would remove the expired-schedule debug logging, but I don't think that's a big loss really.

I'll look into this.